manta-accounting/src/wallet/signer.rs

```diff
@@ -630,63 +633,69 @@ where
-    /// validate the utxo.
+    /// Finds the next viewing key that can decrypt the `encrypted_note` from the `view_key_table`.
     #[inline]
-    fn insert_next_item(
-        &mut self,
+    fn find_next_key<'h>(
+        view_key_table: &mut ViewKeyTable<'h, C::HierarchicalKeyDerivationScheme>,
         parameters: &Parameters<C>,
         with_recovery: bool,
-        utxo: Utxo<C>,
         encrypted_note: EncryptedNote<C>,
+    ) -> Option<ViewKeySelection<C::HierarchicalKeyDerivationScheme, Note<C>>> {
+        let mut finder = DecryptedMessage::find(encrypted_note);
+        view_key_table
+            .find_index_with_maybe_gap(with_recovery, move |k| {
+                finder.decrypt(&parameters.note_encryption_scheme, k)
+            })
+            .map(|selection| selection.map(|item| item.plaintext))
+    }
+
+    /// Inserts the new `utxo`-`note` pair into the `utxo_accumulator` adding the spendable amount
+    /// to `assets` if there is no void number to match it.
+    #[inline]
+    fn insert_next_item(
+        utxo_accumulator: &mut C::UtxoAccumulator,
+        assets: &mut C::AssetMap,
+        parameters: &Parameters<C>,
+        utxo: Utxo<C>,
+        selection: ViewKeySelection<C::HierarchicalKeyDerivationScheme, Note<C>>,
         void_numbers: &mut Vec<VoidNumber<C>>,
         deposit: &mut Vec<Asset>,
-    ) -> Result<(), SyncError<C::Checkpoint>> {
-        let mut finder = DecryptedMessage::find(encrypted_note);
-        if let Some(ViewKeySelection {
+    ) {
+        let ViewKeySelection {
             index,
             keypair,
-            item,
-        }) = self
-            .accounts
-            .get_mut_default()
-            .find_index_with_maybe_gap(with_recovery, |k| {
-                finder.decrypt(&parameters.note_encryption_scheme, k)
-            })
-        {
-            let Note {
+            item: Note {
                 ephemeral_secret_key,
                 asset,
-            } = item.plaintext;
-            if let Some(void_number) =
-                parameters.check_full_asset(&keypair.spend, &ephemeral_secret_key, &asset, &utxo)
-            {
-                if let Some(index) = void_numbers.iter().position(move |v| v == &void_number) {
-                    void_numbers.remove(index);
-                } else {
-                    self.utxo_accumulator.insert(&utxo);
-                    self.assets.insert((index, ephemeral_secret_key), asset);
-                    if !asset.is_zero() {
-                        deposit.push(asset);
-                    }
-                    return Ok(());
+            },
+        } = selection;
+        if let Some(void_number) =
+            parameters.check_full_asset(&keypair.spend, &ephemeral_secret_key, &asset, &utxo)
+        {
+            if let Some(index) = void_numbers.iter().position(move |v| v == &void_number) {
+                void_numbers.remove(index);
```

Comment:
Why do we have `void_numbers.remove(index)`? `void_number` should be an append-only vector?

Reply:
`void_numbers` is the set of void numbers during this synchronization round. Once we find the pre-image for a void number in this round, we can remove it from the set since it won't be matched again.